server/commands/list_objects.go

```diff
@@ -0,0 +1,279 @@
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.ListObjectsDeadline != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, &PerformChecksInput{
+			storeID:     req.StoreId,
+			authModelID: req.AuthorizationModelId,
+			objectType:  req.Type,
+			relation:    req.Relation,
+			user:        req.User,
+			ctxTuples:   req.ContextualTuples,
+		}, resultsChan, errChan, resolvedChan)
+	}()
+
+	objectIDs := make([]string, 0)
+
+	select {
+	case <-timeoutCtx.Done():
+		// we give them what we have so far
+		for objectID := range resultsChan {
+			objectIDs = append(objectIDs, objectID)
+		}
+
+		partialResults.Add(timeoutCtx, int64(len(objectIDs)))
+
+		return &openfgapb.ListObjectsResponse{
+			ObjectIds: objectIDs,
+		}, nil
+	case <-resolvedChan:
+		for objectID := range resultsChan {
+			objectIDs = append(objectIDs, objectID)
+		}
+
+		allResults.Add(timeoutCtx, int64(len(objectIDs)))
+
+		return &openfgapb.ListObjectsResponse{
+			ObjectIds: objectIDs,
+		}, nil
+	case genericError := <-errChan:
+		return nil, genericError
+	}
+}
+
+func (q *ListObjectsQuery) ExecuteStreamed(ctx context.Context, req *openfgapb.StreamedListObjectsRequest, srv openfgapb.OpenFGAService_StreamedListObjectsServer) error {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.ListObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.ListObjectsDeadline != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, &PerformChecksInput{
+			storeID:     req.StoreId,
+			authModelID: req.AuthorizationModelId,
+			objectType:  req.Type,
+			relation:    req.Relation,
+			user:        req.User,
+			ctxTuples:   req.ContextualTuples,
+		}, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return nil
+		case objectID, ok := <-resultsChan:
+			if !ok {
+				return nil //channel was closed
+			}
+			if err := srv.Send(&openfgapb.StreamedListObjectsResponse{
+				ObjectId: objectID,
+			}); err != nil {
+				return serverErrors.NewInternalError("", err)
+			}
+		case genericError, ok := <-errChan:
+			if ok {
+				return serverErrors.NewInternalError("", genericError)
+			}
+		}
+	}
+}
+
+func (q *ListObjectsQuery) getUniqueObjects(ctx context.Context, storeID, targetObjectType string, ctxTuples *openfgapb.ContextualTupleKeys) (storage.ObjectIterator, error) {
+	iter, err := q.Datastore.ListObjectsByType(ctx, storage.ListObjectsFilter{
+		StoreID:    storeID,
+		ObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	return iter, nil
```

Comment:
Shouldn't this method return an iterator that yields unique objects by combining the iterator returned from the database and the contextual tuples?

It would be nice if the `ObjectIterator` returned here iterates over both so that iteration over the two different relationship tuple streams doesn't have to happen independently in the caller.

Reply:
I had thought of two ways we could do this:

1. Have the storage implementations append the contextual tuples to the iterator they return. This feels like a violation of SRP, the storage implementations shouldn't care about the contextual tuples
2. Add a method in the `ObjectIterator` interface to append the contextual tuples. This also feels wrong... iterator interfaces i've seen only allow you to return elements, not add to them

but I didn't like either, so that's why I kept it as is

Reply:
Can I suggest that we create a `CombinedIterator` in the storage package that takes two iterators and combines them into a single iterator then? We could use this in multiple places. I may recommend we introduce generics a bit to help us make this more elegant. Something like:

```
package storage

type Iterator[T any] interface {
    Next() (T, error)
    Stop()
}

var tupleIterator Iterator[*openfgapb.Tuple]
var objectIterator Iterator[*openfgapb.Object] // see https://github.com/openfga/api/pull/24

func CombinedIterator[T any](iter1, iter2 Iterator[T]) Iterator[T] {
    ...
}

objIter := CombinedIterator[*openfgapb.Object](dbIter, ctxTupleIter)
tupleIter := CombinedIterator[*openfgapb.Tuple](dbIter, ctxTupleIter)
```

I don't think creating a combined iterator at the storage layer is breaking any SRP principles. It's just abstracting away the idea that the results from the database are augmented with the context passed in.

We now have code merging contextual info with database results in multiple places, and that's what I think could be improved.

If we want to do this work in a subsequent PR that's fine, but I think it should be done. If we're not going to do it here to make this code more streamlined and elegant, then let's at least add an issue for this work and follow up on it soon 👍 .

Reply:
I agree that it's a good idea, I hadn't thought of a combined iterator. Let's do it in a separate PR